File: leetcode/two-sum/two-sum.cpp
#include<cstdio>
#include<iostream>
#include<set>
#include<map>
#include<cstring>
#include<string>
#include<algorithm>
#include<cmath>
#include<vector>
#include<stack>
#include<queue>
#include<cctype>
#define LL long long int
#define INF 0x3f3f3f3f
#define EPS 1.0e-8
#define RPE(i,s,e) for(int i = s; i < (int)(e); i++)
#define FOR(i,c) for(__typeof((c).begin()) i = (c).begin(); i != (c).end(); ++i)
#define MEM(c,i) memset(c, i, sizeof(c)) // i爲0或者-1
#define QC(q) while(!q.empty())q.pop()
#define IN(x,low,high) (x>low&&x<high)//low<x<high
#ifdef ONLINE_JUDGE  //判断是不是OJ系统上编译
#define debug(format, ...)  0
#else
char _red_buff[65535];
// {debug,     "\e[0;38m" },
// {info,      "\e[1;37m" },
// {notice,    "\e[1;36m" },
// {warning,   "\e[1;33m" },
// {error,     "\e[1;31m" },
// {critical,  "\e[1;35m" },
// {alert,     "\e[1;44m" },
// {emergency, "\e[1;41m" }
#define debug(format, ...) strcpy(_red_buff, "\e[1;36m");strcat(_red_buff, format);strcat(_red_buff, "\e[0m\r\n");printf(_red_buff, ##__VA_ARGS__)
#endif
using namespace std;
// Definition for singly-linked list with a random pointer.

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int,int> hash;
        for (int i=0; i<nums.size(); i++) {
            int val = target - nums[i];
            auto iter = hash.find(val);
            if (iter == hash.end()) {
                // debug("%d %d\n", i, nums[i]);
                hash.insert(pair<int,int>(nums[i], i));
            } else {
                // debug("%d %d\n", i+1, 1+iter->second);
                return {1+iter->second, i+1};
            }            
        }
    }
};
int main()
{
    Solution s;
    vector<int> nums = {-3,4,3,90};
    s.twoSum(nums, 0);
    return 0;
}

File: leetcode/two-sum/two-sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	hash := make(map[int] int)	
	for i1, v1 := range nums {
		v2 := target - v1
		i2, ok := hash[v2]
		if ok {
			return []int{i2, i1}
		} else {
			hash[v1] = i1
		}
	}
	return []int{};
}
func twoSum0(nums []int, target int) []int {
	ret := []int{}
	len := len(nums)
	for i:=0; i<len; i++ {
		for j:=i+1; j<len; j++ {
			if nums[i]+nums[j] == target {
				ret = []int{i,j}
			}
		}
	}
	return ret;
}
func main() {
	data := []int{2, 7, 11, 15}
	ret := twoSum(data, 9)
	fmt.Printf("%v\n", ret)
}